crates/deltalake-core/src/logstore/mod.rs

```diff
@@ -16,19 +18,71 @@ use crate::{
     kernel::Action,
     operations::transaction::TransactionError,
     protocol::{get_last_checkpoint, ProtocolError},
-    storage::{commit_uri_from_version, config::StorageOptions},
+    storage::{commit_uri_from_version, ObjectStoreRef, StorageOptions},
     DeltaTableError,
 };
 use bytes::Bytes;
-use log::debug;
+use log::*;
 use object_store::{path::Path, Error as ObjectStoreError, ObjectStore};
 
 #[cfg(feature = "datafusion")]
 use datafusion::datasource::object_store::ObjectStoreUrl;
 
 pub mod default_logstore;
-#[cfg(any(feature = "s3", feature = "s3-native-tls"))]
-pub mod s3;
+
+/// Trait for generating [LogStore] implementations
+pub trait LogStoreFactory: Send + Sync {
+    /// Create a new [LogStore]
+    fn with_options(
+        &self,
+        store: ObjectStoreRef,
+        location: &Url,
+        options: &StorageOptions,
+    ) -> DeltaResult<Arc<dyn LogStore>> {
+        Ok(default_logstore(store, location, options))
+    }
+}
+
+/// Return the [DefaultLogStore] implementation with the provided configuration options
+pub fn default_logstore(
+    store: ObjectStoreRef,
+    location: &Url,
+    options: &StorageOptions,
+) -> Arc<dyn LogStore> {
+    Arc::new(default_logstore::DefaultLogStore::new(
+        store,
+        LogStoreConfig {
+            location: location.clone(),
+            options: options.clone(),
+        },
+    ))
+}
+
+#[derive(Clone, Debug, Default)]
+struct DefaultLogStoreFactory {}
+impl LogStoreFactory for DefaultLogStoreFactory {}
+
+/// Registry of [LogStoreFactory] instances
+pub type FactoryRegistry = Arc<DashMap<Url, Arc<dyn LogStoreFactory>>>;
```

Comment:
Just for my own understanding, what does DashMap solve here instead of a normal hashmap? I read the docs but I don't grasp what it's purpose is here : P

Reply:
Dashmap supports lock-free concurrency, so it can be used across threads in the process. HashMap would need to be put into a `Mutex<T>` first in order to do the same.